### ISSUE TITLE
The future of time

### ISSUE THREAD

**Issue by d-chambers** [obspy/obspy#2072]
No, this is not the title of a book written by Stephen Hawking. This issue aims to outline a proposal for future developments to the `UTCDateTime` object and to focus discussion on the topic that has been taking place in a few other threads.

Here are the problems I see, as well as a proposal for fixing the issues. I may be overlooking something so please don't hesitate to disagree. 

# The problem

1. UTCDateTime instances cannot (and should not) be used as keys in a dictionary or members in a set because they are mutable (ie they cannot be safely hashed).

2. UTCDateTime fuzzy comparisons still cause some strange behavior in sorting, min/max, etc. despite some improvements made in #2049. Some of the oddities outlined in #1765 still exists, eg the second half of [this example](https://github.com/obspy/obspy/issues/1765#issuecomment-297705550).

3. Internal representation of nanoseconds results in storing larger numbers than necessary (though the used memory might be the same) for precisions less than 9, and precisions greater than 9 are not possible.

# Proposed changes to `UTCDateTime`

## 1. Make UTCDateTime immutable
There has already been some discussion on #568, #1452, and probably others. Immutability will allow the safe hashing `UTCDateTime` instances as well as protect against a subtle class of bugs (eg unintended mutations). 

This would, of course, mean that instance attributes would no longer be settable. So rather than doing something like this:
```python
import obspy
utc = obspy.UTCDateTime()
utc.year = 1989
utc.month = 5
utc.day = 4
utc.hour = 12
utc.precision  = 3
```  
The user would be required to do this:
```python
import obspy
utc = obspy.UTCDateTime(year=1989, month=5, day=4, hour=12, precision=3)
```

## 2. Use the smallest possible int to represent time

Upon initialization, `UTCDateTime` would round the inputs to the nearest unit according to precision and only store the smallest possible integer to represent the desired precision. For example, for the default  precision of 6, the object would store an int representing microseconds. If precision was set to 0 it would store an int representing seconds, and so on. 

Arguably, if a precision is specified any significant figures exceeding the specified precision are just noise anyway, no reason to store them.  

Moreover, according to [this](https://docs.python.org/2/library/stdtypes.html#numeric-types-int-float-long-complex), longs have no limit on precision, at least until memory is full, so an arbitrary precision could be represented as efficiently as possible by `UTCDateTime`. 

## 3. Raise an error when UTCDateTime objects of different precisions are compared 

This is undefined behavior in many cases and currently raises a warning.  With a different internal representation of time for different precisions this would be especially important. 

## 4. Remove “fuzzy” comparisons

If modifications 2 and 3 are carried out then it would make sense to remove all “fuzzy” comparison behavior. Flat comparisons of the integer representation between `UTCDateTime` objects of the same precision would result in the expected behavior and solve the issues mentioned in point 2. It would also remove the need to round each time a comparison is made, making `UTCDateTime` operations more efficient. 

## 5. Set hash

A tuple of (time_int, precision) would then be a safe (immutable) value to hash on. It may also allow UTCDateTime objects of different precisions to exist in the same dict/set (I need to think about this more).  

# Migration path

There would be two backward incompatible changes, immutability and disallowing comparisons of `UTCDateTime` instances of different precisions. In order to give users plenty of time to adapt, we could start raising a depreciation warning whenever an attribute is set outside the `__init__` method. We could also modify the warning issued when comparing UTCDateTime objects of different precisions to include a message that such operations will raise an error in the future. 

The depreciation warnings would start in obspy 1.2.0, and the changes in UTCDateTime would be implemented in version 1.3.0 or 1.4.0. 

# Alternative 1

We could also leave the nanosecond representation of time as it is, but still round inputs according to precision. This would allow for the removal of fuzzy comparisons, solving issue 2. Then the hash method could simply use the nanosecond int rather than a tuple. Dicts/sets containing UTCDateTime objects with differing precisions, however, would require some more thought. 

This would not solve issue 3, although in practice it probably isn't much of an issue anyway.

By the way, do we have a formal procedure for making OEPs (obspy enhancement proposals) like this one?

**Comment by krischer** [comment]
I agree with everything you wrote there including the migration path. It is also a lot cleaner than the approach you outlined as "Alternative 1" which feel more like bandages to me.

I don't think we have to worry about breaking people's codes when they can no longer compare datetimes with different precisions - most people are not even aware that this is possible. The immutability will hit more people but I still think that this is really required and it should have been immutable in the first place.

I also like the idea in number 2. I'm not sure if there any practical applications for having greater than nanosecond precision but its easy to do in Python so why not.

In term of implementation it should be most practical to add the deprecation warning right away and then implement the rest in one go after `1.2.0` has been released.

> By the way, do we have a formal procedure for making OEPs (obspy enhancement proposals) like this one?

We thought about this but so far we figured we are small enough that most things can just be resolved in Github issues and in the chatroom. Also most proposed changes never really were controversial. If this at one point becomes necessary we should definitely use a more formal structure but until that point I think it adds too much overhead.

**Comment by megies** [comment]
I'm not sure about proposed change 2. While it might not be ideal to use integer nanosecond in some cases, it's still good to have a fixed and simple approach. From my experience that helps in other places where e.g. UTCDateTime objects ought to be serialized (see e.g. #2045). If the internal representation is less rigid and more complex than this also adds complexity for other code down the line.

**Comment by krischer** [comment]
It would not really change anything regarding serialization. You could still serialize as nanosecond timestamps and if we do it right everything you did in #2045 would keep working without requiring any changes.

**Comment by d-chambers** [comment]
> We thought about this but so far we figured we are small enough that most things can just be resolved in Github issues and in the chatroom. 

Maybe we ought to add a "proposal" label in that case?

**Comment by d-chambers** [comment]
Whether we use nanoseconds or a precision dependent representation the work of issuing depreciation warnings in the two cases outlined above will still need to be done. We will then have until the 1.2.0 release to discuss more before any other changes are made.

I will probably wait a few days to give more people a chance to see this issue and weigh in if they want to. If there are no objections at that point I will create a PR to add the warnings.

**Comment by trichter** [comment]
I strongly support to make UTCDateTime immutable and hashable. I cannot assess the rest (2.-4.) properly ... but sounds reasonable.

**Comment by ThomasLecocq** [comment]
I'm going through some old-but-good discussions and evaluate (with some help from a friend) what still applies today: 

claudy answer:

Checking in on this 7 years later — worth flagging that the ground has shifted, because the switch to integer-nanosecond internal storage (`_ns`) has since landed, and that quietly resolves the premise behind a good chunk of the proposal.

The original motivation for fuzzy comparisons was float round-off. That's essentially gone now:

```python
>>> a = UTCDateTime(0)
>>> for _ in range(3): a = a + 0.1
>>> a._ns, UTCDateTime(0.3)._ns
(300000000, 300000000)                    # exact; == is True with no fuzz
>>> UTCDateTime(1.1 + 2.2)._ns == UTCDateTime(3.3)._ns
True
```

Float seconds are rounded to ns at construction, so exact comparison already "just works" for the classic cases. The only thing fuzzy `==` still does is treat differences *below the stated precision* as equal (at precision 6, `0ns == 1ns` → `True`) — and that's exactly what makes it non-transitive and breaks sorting/min/max (#1765, still a release blocker).

Going through the five proposals with today's eyes:

- **(2) smallest-possible int representation** — correctly dropped, keep it dropped. Fixed int64 ns is simple, fast, serializable; "smallest int" saves nothing real and sub-ns precision has ~no use here.

- **(4) remove fuzzy comparisons** — the one that's genuinely a bug rather than a nicety, and its original justification is now obsolete. IMO the fix isn't "remove comparison logic" but "make it deterministic": compare on `_ns`, and when precisions differ round both to the coarser precision and then compare exactly. Transitive, sortable, matches `numpy.datetime64` / `pandas.Timestamp`. Closes #1765.

- **(5) hashable** — still wanted (#568), but downstream of (4): you can't have a valid `__hash__` while `==` is non-transitive. Once `==` is exact on `_ns`, hashing is basically `hash(self._ns)` (or `(_ns, precision)`), and #568 becomes a two-liner.

- **(1) immutability** — mostly earns its keep as the thing that makes hashing *safe* (don't hash a mutable object). On its own it's a "should've been this way" cleanup with a real if small migration cost (people do set `.precision`, our own code might mutate). Worth bundling with (5), not urgent alone. FWIW the planned deprecation-warning-on-mutation was never even added in 7 years, which says something about how much pain it's actually causing.

- **(3) raise on different-precision comparison** — this is the one I'd actively *not* do. A `==` that raises is a Python anti-pattern; it breaks `in`, dict membership, `sorted`, and all sorts of incidental comparisons in surprising places. The stale "will raise an Exception in a future version" warning should be reconsidered rather than fulfilled — the clean answer is to *define* cross-precision semantics (round to coarser, compare exactly), which falls out of (4) for free.

So the whole thing basically collapses to one change worth doing: **make comparison deterministic on integer ns**. That fixes #1765, makes hashing well-defined (so #568 follows), and dissolves the need for (3). (1) is a reasonable optional companion for safe hashing; (2) stays dead.

Happy to prototype the comparison change to see how much of the test suite actually leans on the current fuzzy behaviour.
